### PR TITLE
Initial API requirements

### PR DESCRIPTION
**Pull request by samkearney**

### CODE REVIEW

File: include/rdmnet/cpp/controller.h [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+                                         const etcpal::SockAddr& static_broker_addr = etcpal::SockAddr{});
+  etcpal::Result RemoveScope(ScopeHandle handle);
+
+  etcpal::Expected<uint32_t> SendRdmCommand(ScopeHandle scope, const LocalRdmCommand& cmd);
+  etcpal::Result SendRdmResponse(ScopeHandle scope, const LocalRdmResponse& resp);
+  etcpal::Result SendLlrpResponse(const LlrpLocalRdmResponse& resp);
+  etcpal::Result RequestClientList(ScopeHandle scope);
+
+  ControllerHandle handle() const;
+  const etcpal::Uuid& cid() const;
```
Comment (samkearney) [inline include/rdmnet/cpp/controller.h]: Add ScopeHandle to string (or ScopeHandle to scope info)

File: include/rdmnet/cpp/controller.h [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+};
+
+/// \brief An instance of RDMnet Controller functionality.
+/// \ingroup rdmnet_controller_cpp
+class Controller
+{
+public:
+  etcpal::Result Startup(const etcpal::Uuid& cid, ControllerNotify& notify, RdmUid& uid = RdmUid{},
+                         const std::string& search_domain = std::string{},
+                         const std::vector<LlrpNetintId>& llrp_netints = std::vector<LlrpNetintId>{});
```
Comment (samkearney) [inline include/rdmnet/cpp/controller.h]: Generalize llrp_netints as mcast_netints

File: include/rdmnet/cpp/controller.h [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+/// @{
+
+using ControllerHandle = rdmnet_controller_t;
+using ScopeHandle = rdmnet_client_scope_t;
+
+/// @}
+
+/// \brief A base class for a class that receives notification callbacks from a controller.
+/// \ingroup rdmnet_controller_cpp
+class ControllerNotify
```
Comment (samkearney) [inline include/rdmnet/cpp/controller.h]: Find better naming for this interface. Name should be noun (e.g. ControllerNotifyHandler or something) and function names should be verbs (e.g. starting with Handle...)

**Comment by samkearney** [comment]
@nwagner 

> any implications to how the different memory styles change?

In addition to making the APIs nicer, it also makes memory management easier and improves cache locality. The library has two memory usage paradigms:
1. malloc() is used (`RDMNET_DYNAMIC_MEM=1`) - messages are built as the TCP stream comes in off the wire, with realloc() called if necessary to add more items to an array, the equivalent of push_back() to a vector - the message is delivered to a callback once complete.
2. Static memory is used (`RDMNET_DYNAMIC_MEM=0`), _and_ messages are only parsed and delivered from a single thread (`RDMNET_ALLOW_EXTERNALLY_MANAGED_SOCKETS=0`) - static storage is defined large enough to hold X messages, which are delivered to a callback from a thread, then that storage is reused on the next iteration. Much simpler than the current method actually, and nicer for cache.

> I'm a little uneasy about the change from having the compiler calculate size totals to having it in a comment.

Why?

**Comment by nwagner** [comment]
@samkearney, the uneasiness is a very minor one -- what if there was a math error in your comment? I'm fine with you keeping it as is.

**Comment by samkearney** [comment]
@nwagner I made that change because not only does it look cleaner at the site of the definition, it also looks much nicer when someone hovers over a use of the macro when they have some sort of Intellisense available. They instantly know the total without having to do the addition themselves.

I'm confident in the ability of unit tests and code review to catch any potential math errors.

**Comment by nwagner** [comment]
Your Intellisense point is well taken.  Thanks!

**Pull request closed without merging**